File: Example/PayTPV Objc Example/ModelStore.m
//
//  ModelStore.m
//  PAYTPV Objc Example
//
//  Created by Mihail Cristian Dumitru on 8/18/17.
//  Copyright © 2017 PAYTPV. All rights reserved.
//

#import "ModelStore.h"

@interface ModelStore()

@end

@implementation ModelStore

static NSString *kCardStore = @"kCardStore";
static NSString *kPurchaseStore = @"kPurchaseStore";

#pragma mark - Cards

+ (nonnull NSArray *)cards {
    NSData *cardsData = [[NSUserDefaults standardUserDefaults] dataForKey:kCardStore];
    if (cardsData == nil) {
        return [NSArray new];
    }
    NSArray *cards = [NSKeyedUnarchiver unarchiveObjectWithData:cardsData];
    if (cards == nil) {
        return [NSArray new];
    }
    return cards;
}

+ (void)saveCards:(NSArray *)cards {
    NSData *cardsData = [NSKeyedArchiver archivedDataWithRootObject:cards];
    [[NSUserDefaults standardUserDefaults] setObject:cardsData forKey:kCardStore];
}

+ (void)cards:(void (^ __nullable)(NSArray * _Nonnull cards))completion {
    if (completion) {
        completion([self cards]);
    }
}

+ (void)addCard:(Card *)card
     completion:(void (^ __nullable)(void))completion {
    NSMutableArray *cards = [NSMutableArray arrayWithArray:[self cards]];
    [cards addObject:card];
    [self saveCards:cards];
    if (completion) {
        completion();
    }
}

+ (void)removeCard:(Card *)card
        completion:(void (^ __nullable)(void))completion {
    NSMutableArray *cards = [NSMutableArray arrayWithArray:[self cards]];
    [cards removeObject:card];
    [self saveCards:cards];
    if (completion) {
        completion();
    }
}

#pragma mark - Purchases

+ (nonnull NSArray *)purchases {
    NSData *purchasesData = [[NSUserDefaults standardUserDefaults] dataForKey:kPurchaseStore];
    if (purchasesData == nil) {
        return [NSArray new];
    }
    NSArray *purchases = [NSKeyedUnarchiver unarchiveObjectWithData:purchasesData];
    if (purchases == nil) {
        return [NSArray new];
    }
    return purchases;
}

+ (void)savePurchases:(NSArray *)purchases {
    NSData *purchasesData = [NSKeyedArchiver archivedDataWithRootObject:purchases];
    [[NSUserDefaults standardUserDefaults] setObject:purchasesData forKey:kPurchaseStore];
}

+ (void)purchases:(void (^ __nullable)(NSArray * _Nonnull purchases))completion {
    if (completion) {
        completion([self purchases]);
    }
}

+ (void)addPurchase:(Purchase *)purchase
         completion:(void (^ __nullable)(void))completion {
    NSMutableArray *purchases = [NSMutableArray arrayWithArray:[self purchases]];
    [purchases addObject:purchase];
    [self savePurchases:purchases];
    if (completion) {
        completion();
    }
}

+ (void)removePurchase:(Purchase *)purchase
            completion:(void (^ __nullable)(void))completion {
    NSMutableArray *purchases = [NSMutableArray arrayWithArray:[self purchases]];
    [purchases removeObject:purchase];
    [self savePurchases:purchases];
    if (completion) {
        completion();
    }
}

@end

File: Example/PayTPV Swift Example/ModelStore.swift
//
//  CardStore.swift
//  PAYTPV Swift Example
//
//  Created by Mihail Cristian Dumitru on 8/18/17.
//  Copyright © 2017 PAYTPV. All rights reserved.
//

import UIKit

/// In a real production app, this should be the api connected to your backend which you can use to send the tokenized user data from PAYTPV. For the purpose of the demo, we just store it in UserDefaults
class ModelStore {

    private static let kCardStore = "kCardStore"
    private static let kPurchaseStore = "kPurchaseStore"

    private static var cards: [Card] = {
        guard let cardsData = UserDefaults.standard.data(forKey: kCardStore) else {
            return []
        }
        guard let cards = NSKeyedUnarchiver.unarchiveObject(with: cardsData) as? [Card] else {
            return []
        }
        return cards
    }()

    private static var purchases: [Purchase] = {
        guard let purchasesData = UserDefaults.standard.data(forKey: kPurchaseStore) else {
            return []
        }
        guard let purchases = NSKeyedUnarchiver.unarchiveObject(with: purchasesData) as? [Purchase] else {
            return []
        }
        return purchases
    }()

    private static func save() {
        let cardsData = NSKeyedArchiver.archivedData(withRootObject: self.cards)
        UserDefaults.standard.set(cardsData, forKey: kCardStore)
    }

    public static func getCards(completion: (([Card]) -> Void)? = nil) {
        completion?(cards)
    }

    public static func add(card: Card, completion: (() -> Void)? = nil) {
        self.cards.append(card)
        self.save()
        completion?()
    }

    public static func remove(card: Card, completion: (() -> Void)? = nil) {
        if let index = self.cards.index(where: { $0 == card }) {
            self.cards.remove(at: index)
        }
        self.save()
        completion?()
    }

    public static func getPurchases(completion: (([Purchase]) -> Void)? = nil) {
        completion?(purchases)
    }

    public static func add(purchase: Purchase, completion: (() -> Void)? = nil) {
        self.purchases.append(purchase)
        self.save()
        completion?()
    }

    public static func remove(purchase: Purchase, completion: (() -> Void)? = nil) {
        if let index = self.purchases.index(where: { $0 == purchase }) {
            self.purchases.remove(at: index)
        }
        self.save()
        completion?()
    }

}

